proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+<interactive>:1:10: error:
+    • Couldn't match type ‘Bool -> Bool’ with ‘forall a. a -> a’
+      Expected type: (Bool -> Bool) -> Bool
+        Actual type: (forall a. a -> a) -> Bool
+    • In the second argument of ‘choose’, namely ‘g’
+      In the expression: choose f g
+Prelude> :t choose g f
+choose g f :: Bool
+```
+
+There is a simple workaround that works most of the time: eta-expansion (you might need to eta-expand more than once, though). Taking the `choose`, `f`, and `g` functions defined above,in a non-contravariant world we need to eta-expand `f` to make it work: `choose (\x -> f x) g`; now the `x` is assigned type `forall a. a -> a`, which is then instantiated to `Int -> Int` before passing it to `f`.
+
+### The `ContravariantFunctions` extension
+
+After a preliminary assessment of which packages still compile under this new policy, it looks like most of them do not need it. Except one big and important one: `Cabal` (and anything that uses its API). This library often uses functions of the form:
+
+```haskell
+f :: ( (... -> CabalIO a) -> ...) -> ...
+```
+
+where `CabalIO a` is defined as `HasCallStack => IO a`. That means that whereas before we could pass a function `g` as argument, now we have to do something akin to `\... t ... -> g (\... x ... -> t ... x ...)`, which is quite annoying.
+
+We thus propose to add another extension to GHC, called `ContravariantFunctions`, which should be disabled by default. When the extension is enabled, subsumption checking is done using full contravariance of function types (but during quick look contravariance does not enter the game). This means that `RankNTypes` + `ContravariantFunctions` behaves the same as the old `RankNTypes. This has the benefit of asking people to eta-expand their functions, while having a escape way for those libraries which would require too many changes.
+
+### The typing rule for `($)`
+
+Currently, GHC contains a hard-coded typing rule for `($)` (you can find it in [this file](https://gitlab.haskell.org/ghc/ghc/blob/master/compiler/typecheck/TcExpr.hs), looking for `dollarIdKey`) which ensures that `f $ e` works even when `($)` would need to be impredicatively instantiated. The million dollar question is: can we drop this special case from the compiler?
+
+A few versions ago, the answer would have been **no**. The reason was that `($)` was the only function able to operate with both lifted and unlifted kinds. But nowadays, this fact is reflected in its type:
+
+```haskell
+($) ::
+  forall (r :: GHC.Types.RuntimeRep) a (b :: TYPE r).
+  (a -> b) -> a -> b
+```
+
+So the initial answer would be **yes**. This should be taken with a grain of salt, though, since the specification in the draft paper does not consider polymorphic type representations. However, for the usual case of `b` having kind `*`, the rule may be dropped. This also means that other combinators such as `(&)` or `(.)` no longer are second-class with respect to impredicativity.
```

Comment:
I'd like to see, here, an example which works out how `runST $ do {…}` would typecheck without the special rule.

Reply:
By the way, a discussion which is missing is that in current GHC:

```haskell
> import Control.Monad.ST
> runST $ return 0
0
```

Works without extension. If the special rule for `($)` is removed, then the above will require `-XImpredicativeTypes`.

Reply:
Added discussion to the proposal.